File: src/main/java/tarmorn/Eval.java
package tarmorn;

import java.io.IOException;
import java.util.ArrayList;

import tarmorn.data.Triple;
import tarmorn.data.TripleSet;
import tarmorn.eval.HitsAtK;
import tarmorn.eval.ResultSet;
import tarmorn.structure.Rule;

public class Eval {
		
	 
	private static String CONFIG_FILE = "config-eval.properties";
	
	
	/**
	 * Read the top k from the ranking file for MRR computation. If less candidates are available, no error is thrown.
	 */
	public static int TOP_K = 100;
	
	
	
	/**
	 * Path to the file that contains the triple set used for learning the rules.
	 */
	public static String PATH_TRAINING = "";
	
	
	
	
	
	/**
	 * Path to the file that contains the triple set used for to test the rules.
	 */
	public static String PATH_TEST = "";
	
	/**
	 * Path to the file that contains the triple set used for validation purpose.
	 */
	public static String PATH_VALID = "";
	
	/**
	 * Path to the output file where the predictions are stored.
	 */
	public static String PATH_PREDICTIONS = "";
	
	
	
	public static void main(String[] args) throws IOException {
		
		
		
		if (args.length == 1) {
			CONFIG_FILE = args[0];
			System.out.println("reading params from file " + CONFIG_FILE);
		}
		
		Rule.applicationMode();
		Settings.load();
		Settings.REWRITE_REFLEXIV = false;
		
		TripleSet trainingSet =  new TripleSet(PATH_TRAINING);
		TripleSet validationSet = new TripleSet(PATH_VALID);
		TripleSet testSet = new TripleSet(PATH_TEST);
		
		
		String[] values = Apply.getMultiProcessing(PATH_PREDICTIONS);

		HitsAtK hitsAtK = new HitsAtK();
		hitsAtK.addFilterTripleSet(trainingSet);
		hitsAtK.addFilterTripleSet(validationSet);
		hitsAtK.addFilterTripleSet(testSet); 
		
		StringBuilder sb = new StringBuilder();
		if (values.length == 1) {
			ResultSet rs = new ResultSet(PATH_PREDICTIONS, true, TOP_K);
			computeScores(rs, testSet, hitsAtK);
			System.out.println(hitsAtK.getHitsAtK(0) + "   " + hitsAtK.getHitsAtK(2) + "   " + hitsAtK.getHitsAtK(9) + "   " + hitsAtK.getApproxMRR());
		}
		else {
			for (String value : values) {
				
				String rsPath = PATH_PREDICTIONS.replaceFirst("\\|.*\\|", "" + value); 
				ResultSet rs = new ResultSet(rsPath, true,  TOP_K);
				computeScores(rs, testSet, hitsAtK);
				System.out.println(hitsAtK.getHitsAtK(0) + "   " + hitsAtK.getHitsAtK(2) + "   " + hitsAtK.getHitsAtK(9) + "   " + hitsAtK.getApproxMRR());
				sb.append(value + "   " + hitsAtK.getHitsAtK(0) + "   " + hitsAtK.getHitsAtK(9) + "   " + hitsAtK.getApproxMRR() + "\n");
			}
			System.out.println("-----");
			System.out.println(sb);
		}

		
	
	}

	
	private static void computeScores(ResultSet rs, TripleSet gold, HitsAtK hitsAtK) {
		for (Triple t : gold.getTriples()) {		
			ArrayList<String> cand1 = rs.getHeadCandidates(t.toString());
			// String c1 = cand1.size() > 0 ? cand1.get(0) : "-";
			hitsAtK.evaluateHead(cand1, t);
			ArrayList<String> cand2 = rs.getTailCandidates(t.toString());
			// String c2 = cand2.size() > 0 ? cand2.get(0) : "-";
			hitsAtK.evaluateTail(cand2, t);
		}
	}
	
	

		
}

File: src/main/java/tarmorn/Eval.kt
package tarmorn

import tarmorn.Settings.load
import tarmorn.data.TripleSet
import tarmorn.eval.HitsAtK
import tarmorn.eval.ResultSet
import tarmorn.structure.Rule
import java.io.IOException

object Eval {
    private var CONFIG_FILE: String? = "config-eval.properties"


    /**
     * Read the top k from the ranking file for MRR computation. If less candidates are available, no error is thrown.
     */
    var TOP_K: Int = 100


    /**
     * Path to the file that contains the triple set used for learning the rules.
     */
    var PATH_TRAINING: String = ""


    /**
     * Path to the file that contains the triple set used for to test the rules.
     */
    var PATH_TEST: String = ""

    /**
     * Path to the file that contains the triple set used for validation purpose.
     */
    var PATH_VALID: String = ""

    /**
     * Path to the output file where the predictions are stored.
     */
    var PATH_PREDICTIONS: String = ""


    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        if (args.size == 1) {
            CONFIG_FILE = args[0]
            println("reading params from file " + CONFIG_FILE)
        }

        Rule.applicationMode()
        load()
        Settings.REWRITE_REFLEXIV = false

        val trainingSet = TripleSet(PATH_TRAINING)
        val validationSet = TripleSet(PATH_VALID)
        val testSet = TripleSet(PATH_TEST)


        val values = Apply.getMultiProcessing(PATH_PREDICTIONS)

        val hitsAtK = HitsAtK()
        hitsAtK.addFilterTripleSet(trainingSet)
        hitsAtK.addFilterTripleSet(validationSet)
        hitsAtK.addFilterTripleSet(testSet)

        val sb = StringBuilder()
        if (values.size == 1) {
            val rs = ResultSet(PATH_PREDICTIONS, true, TOP_K)
            computeScores(rs, testSet, hitsAtK)
            println(hitsAtK.getHitsAtK(0) + "   " + hitsAtK.getHitsAtK(2) + "   " + hitsAtK.getHitsAtK(9) + "   " + hitsAtK.getApproxMRR())
        } else {
            for (value in values) {
                val rsPath = PATH_PREDICTIONS.replaceFirst("\\|.*\\|".toRegex(), "" + value)
                val rs = ResultSet(rsPath, true, TOP_K)
                computeScores(rs, testSet, hitsAtK)
                println(hitsAtK.getHitsAtK(0) + "   " + hitsAtK.getHitsAtK(2) + "   " + hitsAtK.getHitsAtK(9) + "   " + hitsAtK.getApproxMRR())
                sb.append(value + "   " + hitsAtK.getHitsAtK(0) + "   " + hitsAtK.getHitsAtK(9) + "   " + hitsAtK.getApproxMRR() + "\n")
            }
            println("-----")
            println(sb)
        }
    }


    private fun computeScores(rs: ResultSet, gold: TripleSet, hitsAtK: HitsAtK) {
        for (t in gold.getTriples()) {
            val cand1 = rs.getHeadCandidates(t.toString())
            // String c1 = cand1.size() > 0 ? cand1.get(0) : "-";
            hitsAtK.evaluateHead(cand1, t)
            val cand2 = rs.getTailCandidates(t.toString())
            // String c2 = cand2.size() > 0 ? cand2.get(0) : "-";
            hitsAtK.evaluateTail(cand2, t)
        }
    }
}
